### PR TITLE
Implementation of Country New Month Scripting

### PR DESCRIPTION
**Pull request by MaxMahem**
Implementation of NewMonthCountry scripting. While NewMonthCountry takes a mutable/non-const reference to `Country` it does not expose any non-const methods to the scripting engine, aside from the ability to set scripting values. Instead, new values are returned, and C++ handles any changes to state after scripting exits. Doing it this way required some reworking of the involved methods, but they are probably better for it. Scripts are stored in the `RuleCountry` class, though both classes can take tags.

NewMonthCountry runs before `Country`'s state is changed by NewMonth, but it is passed in the proposed changes to the country's state. The values returned by `NewMonthCountry` are applied to `Country` or the proposed/default changes are applied if there is no change.

No specific constraints are imposed on the scripting return logic. Arbitrary transformations are possible, including ones that are invalid under default rules, such as a country having an alien pact but still providing funding or even providing negative funding. It is assumed that if you want to alter these rules, you mean what you say, so there isn't any point in enforcing existing constraints. This also means that if you want to perform some standard logic via script, you need to reimplement that logic via script, though this is straightforward to do, and I can provide an example script.

Of course, if no script is invoked, then the existing rules apply as standard.

There was no added provisions for getting a RNG generator, one should already be accessible via `GeoscapeGame.getRandomState`. 

Satisfaction was transformed into an enum as just some housekeeping, as its only purpose is to communicate state back to the end-of-month function. If a script wants to use satisfaction for other purposes, it can apply its own tag.

### CODE REVIEW

File: src/Savegame/Country.cpp [hunk 1]
```diff
@@ -333,4 +368,76 @@ bool Country::canBeInfiltrated()
 	return false;
 }
 
+namespace
+{
+
+int checkBoundsAndGet(int index, const std::vector<int>& element)
+{
+	if (index < 0 || index > element.size()) {
+		Log(LOG_WARNING) << "Attempted to get invalid data point.";
```
Comment (Yankes) [inline src/Savegame/Country.cpp:377]: Do not spam logs when error happens

File: src/Savegame/Country.cpp [hunk 2]
```diff
@@ -333,4 +368,76 @@ bool Country::canBeInfiltrated()
...
+		return -1;
+	}
+	return element[index];
+}
+
+}
+
+int Country::getMonthlyActivityAlien(int month) const { return checkBoundsAndGet(month, _activityAlien); }
```
Comment (Yankes) [inline src/Savegame/Country.cpp:385]: This function is not ergonomic for scripts, rewrite them to give "past" months from curret one, like:
```
foo(0) -> _activityAlien.back();
foo(1) -> _activityAlien[_activityAlien.size() - 2]
foo(2) -> _activityAlien[_activityAlien.size() - 3]
```
When you go over beginning you return `0` not `-1`.

File: src/Mod/ModScript.h [hunk 3]
```diff
@@ -238,6 +239,15 @@ class ModScript
 		}
 	};
 
+	////////////////////////////////////////////////////////////
+	//					country script
+	////////////////////////////////////////////////////////////
+	// Output(fundingChange, newSatisfaction, newPact, cancelPact), Input(Country, Save, XComScore, AlienScore)
+	struct NewMonthCountryParser : ScriptParserEvents<ScriptOutputArgs<int&, int&, int&, int&>, Country*, SavedGame*, int, int>
```
Comment (Yankes) [inline src/Mod/ModScript.h:246]: What is point to giving writable access to both game objects?
If it do not have use it should be make read-only.

**Review by Yankes — COMMENTED** [review]
Some comments

File: src/Mod/ModScript.h (hunk 3 above)
Comment (MaxMahem) [reply src/Mod/ModScript.h:246]: `GeoscapeGame.getRandomState` appears to require a PTRE. Likewise, all of the `SetTag` methods appear to require a PTRE. However, while the objects are non-const, neither object actually appears to offer access to any method that can mutate state outside of changing scripting values. `Country` certainly does not.

File: src/Savegame/Country.cpp (hunk 2 above)
Comment (MaxMahem) [reply src/Savegame/Country.cpp:385]: So you would just prefer this to increment backwards? I can do that, though my thinking was to provide an interface that could be easily looped over, in order to collect averages or sums or whatever over a range or portion of a range.

File: src/Savegame/Country.cpp (hunk 1 above)
Comment (MaxMahem) [reply src/Savegame/Country.cpp:377]: ? Getting a value out of range is likely never desired behavior. You would just rather it fail silently? `getMonthsTracked` is exposed so the scripting context can know the size of the tracked data.

File: src/Mod/ModScript.h (hunk 3 above)
Comment (Yankes) [reply src/Mod/ModScript.h:246]: You can pass only copy of random state. This would be beneficial as it will make global random deterministic and independent from all script logic.

And `SetTag` is this need there? as we allow `ptre` them make this function "cemented" in engine. This mean in long run any change to country or related data will be alterable in this place and engine can't reason what state is of `Country`.

This mean you could make country sign pact but script will return that country do not signed any thing, or opposite.

To sum up, what is exactly "end game" for this script? if only to return 4 values, then it should not alter anything else.

File: src/Savegame/Country.cpp (hunk 2 above)
Comment (Yankes) [reply src/Savegame/Country.cpp:385]: But your version need to handle first 12 month in special way, you need multiple `if` to handle like less than 6 months.
If you use offset then you can always count last 6months same way even if you start game or if you are in 10y into game.
Another thing is that if we even expand length of tracked months then all mods will break if they assume 12 max.

File: src/Savegame/Country.cpp (hunk 1 above)
Comment (Yankes) [reply src/Savegame/Country.cpp:377]: if mod is bugged and call multiple times this function then we could flood logs (maybe in this case it would be hard to archive this but this should be rule how we handle messages in scripts).

File: src/Savegame/Country.cpp (hunk 2 above)
Comment (MaxMahem) [reply src/Savegame/Country.cpp:385]: I guess I'm not following you. My thinking was I wanted to allow something like this to be easily possible:
```
var int months;
var int sum 0;

country.getMonthsTracked months;
# we want a 3 month average.
if months gt 3;
  set months 3;
end;

loop var month months;
  var int activity;
  country.getMonthlyActivityAlien activity;
  add sum activity;
end;

div sum months;
# do stuff with the average calculated.
```

I don't really follow the rest of your arguments, but I suppose it doesn't matter. I guess this can be done backward as easily as it is forwards, so I'll make the change.

**Comment by Yankes** [comment]
Small request, I think this feature should consider when you can implement some mod that have non trivial logic that use this new hook.
I think to consider including this to engine you should allow to implement mod that:
"Civilians deaths give penalty from country where attack happens"

File: src/Savegame/Country.cpp (hunk 2 above)
Comment (Yankes) [reply src/Savegame/Country.cpp:385]: Compare this to:
```
var int m;
var int sum 0;

country.getMonthlyActivityAlien m 0;
add sum m;
country.getMonthlyActivityAlien m 1;
add sum m;
country.getMonthlyActivityAlien m 2;
add sum m;
div sum 3;
```
And this will work at day 1 and at day 10000, and your version need handling if you want get only last 3 months.
I think I would even add helper function that `add`:
```
var int sum 0;
country.addMonthlyActivityAlien sum 0;
country.addMonthlyActivityAlien sum 1;
country.addMonthlyActivityAlien sum 2;
div sum 3;
```
You could add loop too, but this would be cherry on top.

**Comment by MaxMahem** [comment]
> Small request, I think this feature should consider when you can implement some mod that have non trivial logic that use this new hook. I think to consider including this to engine you should allow to implement mod that: "Civilians deaths give penalty from country where attack happens"

I'm open to doing this, (probably a change to `DebriefingState`? I think there is already a hook here), though it feels like something for a different pull request.

**Comment by Yankes** [comment]
> > Small request, I think this feature should consider when you can implement some mod that have non trivial logic that use this new hook. I think to consider including this to engine you should allow to implement mod that: "Civilians deaths give penalty from country where attack happens"
> 
> I'm open to doing this, (probably a change to `DebriefingState`? I think there is already a hook here), though it feels like something for a different pull request.

More on kill itself, you do not add any new scripts hooks, only expose country to other scripts and for start for battlestate to have info what country have "privilege" to have this battle on.

File: src/Savegame/Country.cpp (hunk 1 above)
Comment (MaxMahem) [reply src/Savegame/Country.cpp:377]: I mean, if a script is doing something wrong the user/developer really should be informed about that, *especially* if it is wrong in such a way that it ends up flooding a bunch of error messages (this probably means something is *really* wrong). So I guess I just disagree fundamentally here. 

Frankly, I think there is an argument to be made that we should just throw an exception and *die* in this place if a scripting error occurs so that the script writer can definitely be alerted to it. But since there is a reasonable way to recover from this sort of error, I thought recovering and alerting was a better approach.

The only reason to error silently is if the user making an erroneous call was desirable behavior. Say we wanted to let the user easily query all 12 months of data regardless of whether there is data. I could see an argument for this, but then the problem becomes how do we distinguish between a return result that is potentially valid data (like 0) and a return result that indicates an absence of data (currently returning -1, since under base logic none of these values should ever be negative). But if a script's logic is expanded to handle this distinction, it might as well instead expand not to make erroneous calls in the first place.

File: src/Mod/ModScript.h (hunk 3 above)
Comment (Yankes) [reply src/Mod/ModScript.h:246]: > > You can pass only copy of random state. This would be beneficial as it will make global random deterministic and independent from all script logic.
> 
> Is this really beneficial? From the player's point of view, scripting _is_ a part of the game's state, so it should follow the same rules as the rest of the game. If you really want some sort of scriptable rng that doesn't influence the game's state, I would add something that does that to the list of "global" math functions. I'm open to writing that, though I don't see it as beneficial as opposed to following the same rules as everything else.
> 

This have noting with player. This have every thing with engine and debugging and reproducibility. If you use only once global rng all logic after will change its behavior as different value it get for rng.
If you do not touch global rng but use copy, then this script will no way affect other logic in game aside for return values.
This mean script will be near PURE functions.

> > And SetTag is this need there?
> > Well, frankly, my preference would be that script tags _always_ be mutable from a scripting context. But I haven't looked into how hard a change that would be at this point (mark all the scriptvalues mutable? I dunno). If a script can't carry forward some sort of information from invocation to invocation, it really restricts what you can do with them.
> 

As some time ago I said, script _always_ should get `const` objects aside case where whole point of script is change global state or specific object in arbitrary way.

> As for this specific hooks requirement, it's necessary to implement any logic that tracks state outside what `Country` currently enables. For example, I had the idea that a country could "leave" the X-com project due to poor performance but not sign an alien pact. Or perhaps not be in the project, to begin with, but joining the project later. This could be implemented by having an "allied" tag, which is set or unset via some internal scripting logic as appropriate. The possibilities grow as `Country` is exposed in more places, but really it's about keeping options unfettered.
> 

Question if this function should be responsible for updating global state. May be correct solution would be callback that is call at end of day for country? and in current hook you add it only do final decision.

Your current approach is problematic as can't make easy iteration between countries, you have only one chance to make decision and you could not have access to other countries or other country already make decision.

Effective to have best control over this you would need have multiple pass of script hook to correctly handle cross references.

> > This mean in long run any change to country or related data will be alterable in this place and engine can't reason what state is of Country.
> > This mean you could make country sign pact but script will return that country do not signed any thing, or opposite.
> 
> Well, I mean the logic as is can already alter the state of `Country` in more or less arbitrary ways. It can already override forming a pact or not forming a pact, change funding to an arbitrary value, changing satisfaction arbitrarily, ect. While no exposed method mutates state, the script as a whole is already mutable in more or less every way that matters.
> 

but state is change by code after this hook, and this is critical difference. if all logic would be in this script hook, then you would be 100% correct, now we have split responsibility that I would like to avoid.

> And I mean, that was my design intent that the scripting logic be able to write its own rules, so to speak, about `Country` state changes at month end. I initially wrote it in a mutable way because of this but changed it to work in a more "pure" function-like way because some minor parts of the logic were cleaner like this. But it's really a difference in style, not functionality.
> 

Yes, goal is to have arbitrary logic there, but it need have we need have bounds on what is possible (no some thing like `delete country;`) and code after `damageUnit` need run cleanups of game state of all possible changes after script.

> I think, bigger picture, it's probably more productive to handle immutability in a more "object oriented" way. Mutating state being controlled more by access or lack-there-of to methods that can mutate private data. In fact, with scripting, we can do it one better by perhaps _only_ registering "mutable" methods to the specific functions that require them (in their Parser constructor).
>
> But failing that, yeah, I think script tags need to be always writable from the context of a script. That was the main motivating factor in using a PTRE here.

No, again, they should be const as this make all code around them simple and easy to reason ablaut.
We play there long run game, scripts should not be able to corrupt game state, and this need heavy use of `const`.
I simply will not allow situation that to correctly debug game crash you need debug scripts. Using `ptre` only is asking
for more place where script could make holes in logic.

Using `ptre` make too engine code harder to refactor as we need consider all visible side effects that script can relay on.
If we add `ptre` in your new script then we could never change order in with script are called or when are called.

File: src/Savegame/Country.cpp (hunk 2 above)
Comment (MaxMahem) [reply src/Savegame/Country.cpp:385]: Actually, this would fail on the first and second month, because, at that point there are only 1-2 months of data :P

But getting the months in forward or reverse order is a matter of little difference. I'll make the change.

Comment (Yankes) [reply src/Savegame/Country.cpp:385]: > Actually, this would fail on the first and second month, because, at that point there are only 1-2 months of data :P
> 
> But getting the months in forward or reverse order is a matter of little difference. I'll make the change.

How? whole point is that you can call `country.getMonthlyActivityAlien m 10000;` and it should work and set
value `0` to `m`.
This mean it will work for first game months too, of corse average will be 1/3 of only value, but this will be consistent with case when last year have no actions.

File: src/Savegame/Country.cpp (hunk 1 above)
Comment (MaxMahem) [reply src/Savegame/Country.cpp:377]: I get you in general, but this specific case is problematic. Take the example I used earlier. I want to base my logic on the last three months of alien activity, but what if there is only one month of data? If I silently return 0 for the absent months without telling the user that I don't have data for these months, they could average two 0's with the one month of data I have and get an incorrect result. But at the same time, zero is also possibly a valid (and common) result for UFO/XCom activity in a region, so it is not possible to distinguish between "no data" and "we have data, and the value is zero" by returning zero in these cases.

That is to say, "we return the data if we have it, and zero if we don't" is a bad contract if data can potentially also contain zero values, which this data can. Returning an out-of-bounds value (-1) is a slightly better contract. But still, since doing so probably indicates the user is doing something in error, I think it's best to warn them (with an error message) is probably the best behavior.

File: src/Savegame/Country.cpp (hunk 2 above)
Comment (MaxMahem) [reply src/Savegame/Country.cpp:385]: Well, you have to make some decisions about what you want to do in the case when you want to average (say) 3 months of data, but there is only one or two months of data. But in most cases, I imagine using 0 for the missing value is probably not the behavior most frequently wanted. Instead, it generally makes more sense only to average the data present, which argues for returning some value that lets you distinguish between these two cases and/or telling the user that they did something somewhat nonsensical.

File: src/Savegame/Country.cpp (hunk 1 above)
Comment (Yankes) [reply src/Savegame/Country.cpp:377]: If want have this calculate his way then you need add `if` to handle start of game.
I personally dislike calculate average in this way as weights for each data will change from 1 to 1/2, then to 1/3 etc. when months progress.
This mean for starting months average will jump more that after year.
This will could cause problems if you get bad roll in on start of game, if `100` average is "bad" consider sitation that have two events that "cost" `50` each. If both roll in first month (day day 1 and 31) you get `100` average score , if one move one day ahead, then first month will have average `50` and second month will have average `50` too.
In my way of calculating, EVERY month will have close to `33` average, and day when score is calculated will not affect heavy average (first month could have bit less and this is good) .

if we return `-1` lot of naive way to count average will fail and even bugs will be more visible, with `0` some "bugs" will "work" correctly or nor very noticeable.

File: src/Savegame/Country.cpp (hunk 2 above)
Comment (Yankes) [reply src/Savegame/Country.cpp:385]: I think I answer this in other post, to sum up, even if you get data from one month, you need pad values by appending "zeros" to make it behave similar to other month ranges. And if you really need make special handing for first months you can add `if` based on other function you already exposed.

File: src/Savegame/Country.cpp (hunk 1 above)
Comment (MaxMahem) [reply src/Savegame/Country.cpp:377]: I guess we disagree about what is the best default behavior here. But to me, the situation is the user has asked us a bad question. There really isn't any reasonable contract we can form for correct behavior in the face of the question, "Give me data for a month you do not have," since correct behavior depends on how the user wants to handle the absence of data.

So to me, the only appropriate response is to inform the user that they asked us an incorrect question via some sort of error message/return value. I mean, think about how almost all programming languages handle the situation. Almost universally, they will return an error state if you ask for an out-of-range index. I think following this pattern and being upfront and vocal about the limitations of this method is the best way to proceed.

But, I don't want to bikeshed such a trivial issue. Is this a pattern followed elsewhere in the scripting? I can't think of any instances, but my knowledge is hardly encyclopedic.

**Comment by MaxMahem** [comment]
Though I disagree with the methodology, the latest patch implements the requested changes to the monthly getters. They work in reverse order, don't warn on error, and return 0 on error.

I haven't changed the other aspects (ie the constness) as I'm still unclear on what is desired here.

**Comment by Yankes** [comment]
Thanks, for constens, I think better would be dedicated script hook that would run even for each day and it will be able update state of country and other similar functions. and at end of month you only calculate change of funding state.

**Comment by MeridianOXC** [comment]
I will review, amend and submit the "minimum required" and anything beyond the minimum can be discussed afterwards.

**Comment by MeridianOXC** [comment]
Manually merged here.

File: src/Savegame/Country.cpp [hunk 4]
```diff
@@ -84,6 +90,8 @@ YAML::Node Country::save() const
 		node["newPact"] = _newPact;
 	}
 	return node;
+
+	_scriptValues.save(node, &tags);
```
Comment (MeridianOXC) [inline src/Savegame/Country.cpp:94]: Dead code. Should be before return node;

File: src/Savegame/Country.cpp [hunk 5]
```diff
@@ -251,6 +215,77 @@ void Country::newMonth(int xcomTotal, int alienTotal, int pactScore, int average
...
+		return { -lastMonthFunding, Satisfaction::ALIEN_PACT };
+	}
+
+	// canceling a pact and resetting country funding.
+	// note that if a pact is canceled but last month's funding was not 0, this branch is not taken.
+	// This can probably only happen via scripts though, which should handle the issue themselves.
+	if (_cancelPact && lastMonthFunding <= 0)
+	{
+		// using funding appropriate to the country's rules is probably better, but this is the original logic.
+		int averageCountryFunding = save.getCountryFunding() / save.getCountries()->size() / 1000 * 1000;
```
Comment (MeridianOXC) [inline src/Savegame/Country.cpp:260]: this is not equal to original logic, because it can change as the countries are processed. original logic calculated this number only once, at the beginning

File: src/Savegame/Country.cpp [hunk 6]
```diff
@@ -333,4 +368,73 @@ bool Country::canBeInfiltrated()
...
+	countryBinder.addDebugDisplay<&debugDisplayScript>();
+}
+
+/**
+ * Constructor of new month country script parser.
+ * Called every new month for every country.
+ */
+ModScript::NewMonthCountryParser::NewMonthCountryParser(ScriptGlobal* shared, const std::string& name, Mod* mod) : ScriptParserEvents{ shared, name,
+	"fundingChange", "satisfaction", "formPact", "cancelPact",
+	"country", "save", "totalXcomScore", "totalAlienScore" }
```
Comment (MeridianOXC) [inline src/Savegame/Country.cpp:434]: geoscape_game instead of save

File: src/Savegame/Country.cpp [hunk 7]
```diff
@@ -333,4 +368,73 @@ bool Country::canBeInfiltrated()
...
+{
+	Bind<Country> countryBinder = { this };
```
Comment (MeridianOXC) [inline src/Savegame/Country.cpp:436]: BindBase for consistency with other code

File: src/Savegame/Country.h [hunk 8]
```diff
@@ -32,28 +35,35 @@ class RuleCountry;
...
+	/// holds references to user tags. 
+	ScriptValues<Country> _scriptValues;
 public:
 	/// Creates a new country of the specified type.
 	Country(RuleCountry *rules, bool gen = true);
 	/// Cleans up the country.
 	~Country();
 	/// Loads the country from YAML.
-	void load(const YAML::Node& node);
+	void load(const YAML::Node& node, const ScriptGlobal& tags);
```
Comment (MeridianOXC) [inline src/Savegame/Country.h:55]: * instead of & for consistency
